src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```
@@ -706,6 +747,143 @@ static void getFileReaderAndSchema(
+    else
+        return nullptr;
+}
+
+static void updateIncludeTypeIds(
+    DataTypePtr type, const orc::Type * orc_type, bool ignore_case, std::unordered_set<UInt64> & include_typeids)
+{
+    // std::cout << "ch type:" << type->getName() << ", orc_type:" << orc_type->toString() << std::endl;
+
+    /// For primitive types, directly append column id into result
+    if (orc_type->getSubtypeCount() == 0)
+    {
+        include_typeids.insert(orc_type->getColumnId());
+        return;
+    }
+
+    auto non_nullable_type = removeNullable(type);
+    switch (orc_type->getKind())
+    {
+        case orc::LIST: {
+            const auto * array_type = typeid_cast<const DataTypeArray *>(non_nullable_type.get());
+            if (array_type)
+            {
+                updateIncludeTypeIds(
+                    array_type->getNestedType(), orc_type->getSubtype(0), ignore_case, include_typeids);
+            }
+            return;
+        }
+        case orc::MAP: {
+            const auto * map_type = typeid_cast<const DataTypeMap *>(non_nullable_type.get());
+            if (map_type)
+            {
+                updateIncludeTypeIds(map_type->getKeyType(), orc_type->getSubtype(0), ignore_case, include_typeids);
+                updateIncludeTypeIds(map_type->getValueType(), orc_type->getSubtype(1), ignore_case, include_typeids);
+            }
+            return;
+        }
+        case orc::STRUCT: {
+            const auto * tuple_type = typeid_cast<const DataTypeTuple *>(non_nullable_type.get());
+            if (tuple_type)
+            {
+                if (tuple_type->haveExplicitNames())
+                {
+                    const auto & names = tuple_type->getElementNames();
+                    for (size_t tuple_i = 0; tuple_i < names.size(); ++tuple_i)
+                    {
+                        const auto & name = names[tuple_i];
+                        for (size_t struct_i = 0; struct_i < orc_type->getSubtypeCount(); ++struct_i)
```

Consider using `unordered_map` `orc_field_name -> field_index` to avoid nested `for`

---

Also, what will happen if ClickHouse tuple contains field that is not in orc struct? Should we throw an exception if `allow_missing_columns` is false? Or it will be thrown later anyway

---

This wasn't resolved